File: my-folder/0207-course-schedule/solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] prerequisite : prerequisites) {
            int course = prerequisite[0];
            int prerequisiteCourse = prerequisite[1];
            graph.get(course).add(prerequisiteCourse);
        }

        boolean[] visited = new boolean[numCourses];
        boolean[] dfsVisited = new boolean[numCourses];

        for (int i = 0; i < numCourses; i++) {
            if (!visited[i] && hasCycle(graph, visited, dfsVisited, i)) {
                return false;
            }
        }

        return true;
    }

    private boolean hasCycle(List<List<Integer>> graph, boolean[] visited, boolean[] dfsVisited, int course) {
        visited[course] = true;
        dfsVisited[course] = true;

        for (int prerequisiteCourse : graph.get(course)) {
            if (!visited[prerequisiteCourse]) {
                if (hasCycle(graph, visited, dfsVisited, prerequisiteCourse)) {
                    return true;
                }
            } else if (dfsVisited[prerequisiteCourse]) {
                return true;
            }
        }

        dfsVisited[course] = false;
        return false;
    }
}

File: my-folder/0207-course-schedule/solution.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = defaultdict(list)
        for des, src in prerequisites:
            graph[src].append(des)
        
        indegree = [0] * numCourses

        for nodes in graph.values():
            for node in nodes:
                indegree[node] += 1
        
        q = deque([i for i in range(numCourses) if indegree[i] == 0])

        count = 0

        while q:
            node = q.popleft()
            count += 1
            for neighbor in graph[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    q.append(neighbor)
        
        return count == numCourses
